### PR TITLE
Fix clippy warnings

### PR DESCRIPTION
**Pull request by nmurali94**
1. Fix linter warning from clippy about unnecessary borrows - "This expression borrows a reference ... that is immediately dereferenced by the compiler"
2. Replace explicit loop with iterator operations in Header to avoid unnecessary mutation

### CODE REVIEW

**Comment by algesten** [comment]
Hi! Welcome to ureq!

Thank you for the PR, but I think we need to split the work up. Fixing the clippy lint warnings is good, maybe we can reshape this PR to do only that?

Replacing the for-loops with iterators, I'm not so sure about. Typically I find for loops more readable than long iterator chains, and in this specific instance, I think you will end up with worse performance.

The for loop both checks for `:` and illegal chars at the same time.

```rust
        let bytes = self.as_bytes();
        let index = bytes.iter().position(|x| *x == b':').unwrap_or(bytes.len());

        let name_bytes = &bytes[..index];
        let bad = name_bytes.partition_point(is_tchar);
```

This however uses two iterators (`position()` and `partition_point()`) to do the same thing. Reading the docs, I also don't think the use of `partition_point` is correct here.

So to sum it up:

* If you change this PR to addressing the clippy lints only, I'm happy to land that.
* If you want to explore iterators further, please open a separate PR, but beware that I have a mild preference against them, and that I don't want to accept anything that makes performance worse.

**Comment by nmurali94** [comment]
> Fixing the clippy lint warnings is good, maybe we can reshape this PR to do only that?

Will do!

> This however uses two iterators (position() and partition_point()) to do the same thing

My view here was that while we have two iterations here, we're doing half the work in each step so we shouldn't be doing much more work. Plus without a branch check in each step, my intuition says the two iterators can be better optimized by the compiler. But since I don't have anything concrete to show here, I'll defer to your judgement

**Comment by algesten** [comment]
> My view here was that while we have two iterations here, we're doing half the work in each step so we shouldn't be doing much more work. Plus without a branch check in each step, my intuition says the two iterators can be better optimized by the compiler. But since I don't have anything concrete to show here, I'll defer to your judgement

You could be right. I can be convinced by science here ;)

**Review by algesten — APPROVED** [review]
LGTM!
